scripts/check_build_environment.py

```diff
@@ -0,0 +1,145 @@
+        has_problem = True
+    if has_problem:
+        # Only check this if we see an issue above. It may help people with workarounds. However,
+        # it has false positives, so don't check in the normal case.
+        cgo_enabled_env_var = os.environ.get("CGO_ENABLED")
+        if cgo_enabled_env_var != "1":
+            show_problem(
+                f"CGO_ENABLED environment variable is set to {cgo_enabled_env_var}",
+                """Remove no cgo go install, or set CGO_ENABLED=1 in your environment.""",
+            )
+            has_problem = True
+    return has_problem
+
+
+def parse_env_output(s: str):
+    env = {}
+    for l in s.splitlines():
+        if not l:
+            continue
+        k, v = l.split("=", 1)
+        env[k] = v
+    return env
+
+
+def strip_str(s):
+    if hasattr(s, "strip"):
+        return s.strip()
+    return s
+
+
+@check
+def conda_needs_reactivate():
+    has_problem = False
+    if "CONDA_PREFIX" not in os.environ:
+        show_problem(
+            "conda environment not activated",
+            """
+            Run:
+              conda activate katana-dev""",
+        )
+        return True
+    current_env = dict(os.environ)
+    conda_run_env = parse_env_output(
+        subprocess.check_output(["conda", "run", "-p", os.environ["CONDA_PREFIX"], "env"], encoding="UTF-8")
+    )
+    conda_env_name = Path(os.environ["CONDA_PREFIX"]).name
+    for k in set(current_env.keys()) | set(conda_run_env.keys()):
+        if k in ("PS1", "PS2", "SHLVL", "PATH", "CONDA_ROOT", "_"):
+            continue
+        if strip_str(current_env.get(k)) != strip_str(conda_run_env.get(k)):
+            show_problem(
+                f"environment does not match installed conda packages "
+                f"(Environment variable {k} does not match: {current_env.get(k)} != {conda_run_env.get(k)})",
+                f"""
+                Reactivate your environment with:
+                  conda deactivate; conda activate {conda_env_name}
+                """,
+            )
+            has_problem = True
+    return has_problem
+
+
+@check
+def conda_environment_packages_wrong():
+    has_problem = False
+    open_environment_yml = OPEN_ROOT / "conda_recipe" / "environment.yml"
+    enterprise_environment_yml = OPEN_ROOT.parent.parent / "conda_recipe" / "environment.yml"
+    if not open_environment_yml.exists():
+        raise RuntimeError(f"Checkout is missing {open_environment_yml}")
+    for yml in (open_environment_yml, enterprise_environment_yml):
+        if not yml.exists():
+            continue
+
+        with subprocess.Popen(
+            ["conda", "compare", yml], stderr=subprocess.STDOUT, stdout=subprocess.PIPE, encoding="UTF-8"
+        ) as p:
+            output = p.stdout.read()
+            p.wait()
+            if p.returncode != 0:
+                show_problem(f"Conda environment needs update with {yml}\n{output}", CONDA_ENV_UPDATE_INSTRUCTIONS)
+    return has_problem
+
+
+def main():
```

Comment:
Fair. I just think this should be in open. The more we put out tooling in enterprise the less we actually have an open source offering. We are already not looking great on that front.

Is moving dev to open still in the works? It seems pretty important to have our main tooling open for any kind of serious open source offering.

Reply:
> Is moving dev to open still in the works?

No one is actively working on this. Though the blocker here is what exactly the division between the two repos should be.

If you want to at least unify the current scripts we have into a single "dev for non containerized builds" and `dev` can just call `dev_for_non_containerized_builds build/lint/check` as part of its `dev host builb/lint/check` and containerized variants that seems a better path than more scripts.

Reply:
OK. What do you think we should do with this specific tool? I don't want to spend too much time on it and I do think this is important as it would have saved at least two people some pain in the last week.

Reply:
If you want people to regularly run this check, I think we should add a command like `dev host build`, which ensures a configured build directory, activated conda environment if using conda, warns that submodules are out of date, run checks, etc. and does a build.

That command can call whatever scripts are appropriate, perhaps the one in this PR.

As a second-order concern, it would be nice for the external script(s) that `dev host build` calls to map one-to-one with `dev host build` , so that the latter command is simply `exec <external-script> $*` with maybe some more arguments like `--conda-env-file` or whatever.

Bonus points for doing the same with linting.

Reply:
So you are saying we can add this script in a form that is cleaned up to match `dev`. Then we can updated `dev` to call it as a subcommand. Right?

Then I propose this:

* I'll update this script to use `click` and match the style of `dev` as best I can.
* I'll add parameters that `dev` might need to set (like explicit build directory or source directory).
* Then we can integrate this with `dev` as a subcommand either by calling or, down the line, just by moving this code into `dev`.

How does that sound?

I think this would work well as it gets this out there reasonably quickly, but still supports the direction you want to take the custom dev scripts.

Reply:
Sounds good to me.